x-pack/plugin/search-business-rules/src/main/java/org/elasticsearch/xpack/searchbusinessrules/PinnedQueryBuilder.java

```diff
@@ -176,32 +334,39 @@ protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws
     @Override
     protected Query doToQuery(SearchExecutionContext context) throws IOException {
         MappedFieldType idField = context.getFieldType(IdFieldMapper.NAME);
-        if (idField == null) {
+        MappedFieldType indexField = context.getFieldType(IndexFieldMapper.NAME);
+        if (idField == null || indexField == null) {
             return new MatchNoDocsQuery("No mappings");
         }
-        if (this.ids.isEmpty()) {
+        List<Item> items = (documents != null) ? documents : ids.stream().map(id -> new Item(null, id)).collect(Collectors.toList());
+        if (items.isEmpty()) {
             return new CappedScoreQuery(organicQuery.toQuery(context), MAX_ORGANIC_SCORE);
         } else {
-            BooleanQuery.Builder pinnedQueries = new BooleanQuery.Builder();
+            List<Query> pinnedQueries = new ArrayList<>();
 
             // Ensure each pin order using a Boost query with the relevant boost factor
             int minPin = NumericUtils.floatToSortableInt(MAX_ORGANIC_SCORE) + 1;
-            int boostNum = minPin + ids.size();
+            int boostNum = minPin + items.size();
             float lastScore = Float.MAX_VALUE;
-            for (String id : ids) {
+            for (Item item : items) {
                 float pinScore = NumericUtils.sortableIntToFloat(boostNum);
                 assert pinScore < lastScore;
                 lastScore = pinScore;
                 boostNum--;
+                BooleanQuery.Builder documentQueries = new BooleanQuery.Builder();
```

Comment:
Since this method is called on a single index shard, it feels simplest to check check the index right here through `SearchExecutionContext#indexMatches` and omit the item if it doesn't match.

Reply:
Nice—that makes things much simpler. Do I need to handle the case where `pinnedQueries` would be empty or is creating an empty `DisjunctionMaxQuery` ok?

Reply:
I think it's fine to have an empty `DisjunctionMaxQuery`, although it doesn't explicitly optimize this case. Depending on what we decide with allowing duplicates (by making `_index` required or not), we may be also able to stick with a `BooleanQuery` here.